Wikipedia/Code/WMFCVLInfo.m

```diff
@@ -0,0 +1,337 @@
+#import "WMFColumnarCollectionViewLayout.h"
+#import "WMFCVLMetrics.h"
+#import "WMFCVLAttributes.h"
+
+@interface WMFCVLInfo ()
+@property (nonatomic, strong, nonnull) NSMutableArray <WMFCVLColumn *> *columns;
+@property (nonatomic, strong, nonnull) NSMutableArray <WMFCVLSection *> *sections;
+@property (nonatomic, copy, nonnull) WMFCVLMetrics *metrics;
+@end
+
+@implementation WMFCVLInfo
+
+- (instancetype)initWithMetrics:(WMFCVLMetrics *)metrics {
+    self = [super init];
+    if (self) {
+        self.metrics = metrics;
+        [self resetColumns];
+        [self resetSections];
+    }
+    return self;
+}
+
+- (void)resetColumns {
+    self.columns = [NSMutableArray arrayWithCapacity:self.metrics.numberOfColumns];
+    for (NSInteger i = 0; i < self.metrics.numberOfColumns; i++) {
+        WMFCVLColumn *column = [WMFCVLColumn new];
+        column.index = i;
+        column.info = self;
+        [_columns addObject:column];
+    }
+}
+
+- (void)resetSections {
+    self.sections = [NSMutableArray array];
+}
+
+- (id)copyWithZone:(NSZone *)zone {
+    WMFCVLInfo *copy = [[WMFCVLInfo allocWithZone:zone] initWithMetrics:self.metrics];
+    copy.sections = [[NSMutableArray allocWithZone:zone] initWithArray:self.sections copyItems:YES];
+    copy.columns = [[NSMutableArray allocWithZone:zone] initWithArray:self.columns copyItems:YES];
+    for (WMFCVLColumn *column in copy.columns) {
+        column.info = copy;
+        [column enumerateSectionsWithBlock:^(WMFCVLSection * _Nonnull section, NSUInteger idx, BOOL * _Nonnull stop) {
+            section.column = column;
+        }];
+    }
+    copy.boundsSize = self.boundsSize;
+    copy.contentSize = self.contentSize;
+    copy.metrics = self.metrics;
+    return copy;
+}
+
+- (void)enumerateSectionsWithBlock:(nonnull void(^)(WMFCVLSection * _Nonnull section, NSUInteger idx, BOOL * _Nonnull stop))block {
+    [self.sections enumerateObjectsUsingBlock:block];
+}
+
+- (void)enumerateColumnsWithBlock:(nonnull void(^)(WMFCVLColumn * _Nonnull column, NSUInteger idx, BOOL * _Nonnull stop))block {
+    [self.columns enumerateObjectsUsingBlock:block];
+}
+
+- (nullable WMFCVLAttributes *)layoutAttributesForItemAtIndexPath:(NSIndexPath *)indexPath {
+    NSInteger sectionIndex = indexPath.section;
+    
+    if (sectionIndex < 0 || sectionIndex >= self.sections.count) {
+        return nil;
+    }
+    
+    WMFCVLSection *section = self.sections[sectionIndex];
+    NSInteger itemIndex = indexPath.item;
+    if (itemIndex < 0 || itemIndex >= section.items.count) {
+        return nil;
+    }
+    
+    WMFCVLAttributes *attributes = section.items[itemIndex];
+    assert(attributes != nil);
+    return attributes;
+}
+
+- (nullable WMFCVLAttributes *)layoutAttributesForSupplementaryViewOfKind:(NSString *)elementKind atIndexPath:(NSIndexPath *)indexPath {
+    NSInteger sectionIndex = indexPath.section;
+    if (sectionIndex < 0 || sectionIndex >= self.sections.count) {
```

Comment:
Why do we have all the < 0 checks? Being extra careful or did you actually see some?